src/miral/basic_window_manager.cpp

```diff
@@ -89,25 +91,53 @@ miral::BasicWindowManager::Locker::Locker(BasicWindowManager* self) :
         self->workspaces_to_windows.left.erase(workspace);
 }
 
+miral::BasicWindowManager::LockListener::LockListener(
+    std::function<void()> const& on_lock_cb,
+    std::function<void()> const& on_unlock_cb)
+    : on_lock_cb{on_lock_cb},
+      on_unlock_cb{on_unlock_cb}
+{
+}
+
+void miral::BasicWindowManager::LockListener::on_lock()
+{
+    on_lock_cb();
+}
+
+void miral::BasicWindowManager::LockListener::on_unlock()
+{
+    on_unlock_cb();
+}
+
 miral::BasicWindowManager::BasicWindowManager(
     shell::FocusController* focus_controller,
     std::shared_ptr<shell::DisplayLayout> const& display_layout,
     std::shared_ptr<mir::shell::PersistentSurfaceStore> const& persistent_surface_store,
     mir::ObserverRegistrar<mir::graphics::DisplayConfigurationObserver>& display_configuration_observers,
-    WindowManagementPolicyBuilder const& build) :
+    WindowManagementPolicyBuilder const& build,
+    std::shared_ptr<mir::frontend::SurfaceStack> const& surface_stack,
+    std::shared_ptr<mir::scene::SessionLock> const& session_lock) :
     focus_controller(focus_controller),
     display_layout(display_layout),
     persistent_surface_store{persistent_surface_store},
+    surface_stack{surface_stack},
+    session_lock{session_lock},
+    lock_listener{std::make_shared<LockListener>(
+        [this]() { update_application_zones_and_attached_windows(); },
+        [this]() { update_application_zones_and_attached_windows(); }
```

Comment:
Maybe I'm missing something, but why does (un)locking the session require an update?

Reply:
Unlocking requires an update because the exclusion zones for existing windows don't get recalculated automatically after an unlock, meaning that if you "fullscreen" a window, it will appear behind the panels